### ISSUE TITLE
Ipe randomly reorders bitmaps all the time, which is bad for version control

### ISSUE THREAD

**Issue by lorenzhs** [otfried/ipe#356]
Though not a defect, this behaviour is annoying and makes it unnecessarily hard to track ipe files in version control. I recently had the displeasure of merging diverged copies of an ipe file, and that ipe had renumbered the bitmaps certainly didn't help. It also needlessly increases diff sizes. Is there a reason for this behaviour? Similarly, lists of colours, font sizes, opacity levels, etc in ipe styles also get reordered from time to time, for no apparent apparent reason.

**Comment by otfried** [comment]
In principle, the ordering of bitmaps should be deterministic.  However, I have a suspicion why this is not working for you - could you send me a file that exhibits the problem?

**Comment by otfried** [comment]
Also:  we are talking about bitmaps saved as part of an Ipe document, not bitmaps in a style sheet?  It seems the latter code was simply missing the sorting of the bitmaps.

**Comment by lorenzhs** [comment]
I've just emailed you a file that exhibits this behaviour. The bitmaps are saved as part of the Ipe document, *not* in a style sheet. I'm not exactly sure what triggers the reordering, but it happened frequently (but not always) even when I made only trivial changes. For the record, I'm using Ipe 7.2.20 on Linux (Debian unstable).

**Comment by otfried** [comment]
Many thanks for the file - it confirmed my suspicion.  It turns out that while bitmaps are sorted by checksum before saving, my checksum function was buggy.  Instead of a cyclic shift, I had used a linear shift, and so the checksum only depended on the last dozen bytes of the bitmap.  In your file, actually all four bitmaps had the same checksum...

In any case, thanks for making an issue!   I actually noticed the problem of style files changing myself when looking at git diffs, but always was too distracted with the article I was working on to put fixing this on my todo list.

I'm really happy saving will no longer pollute your git log starting in 7.2.22.
